core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/InfrastructureService.java

```diff
@@ -90,7 +96,22 @@ public List<Infrastructure> findAll(Predicate<Infrastructure> filter) {
 
   @PreAuthorize("hasPermission(#infrastructure, 'DELETE')")
   public void delete(Infrastructure infrastructure) {
-    throw new UnsupportedOperationException("Infrastructure deletion not currently supported.");
-  }
+    handlerService.handleDelete(infrastructure);
+    streamBindingView
+      .findAll(sb -> sb.getKey().getInfrastructureName().equals(infrastructure.getKey().getName()))
+      .findAny()
+      .ifPresent(sb -> { throw new IllegalStateException("Infrastructure is used in stream: " + sb.getKey().getStreamKey()); });
+
+    consumerBindingView
+      .findAll(cb -> cb.getKey().getInfrastructureName().equals(infrastructure.getKey().getName()))
+      .findAny()
+      .ifPresent(cb -> { throw new IllegalStateException("Infrastructure is used in consumer binding: " + cb.getKey()); });
 
+    producerBindingView
+      .findAll(pb -> pb.getKey().getInfrastructureName().equals(infrastructure.getKey().getName()))
+      .findAny()
+      .ifPresent(pb -> { throw new IllegalStateException("Infrastructure is used in producer binding: " + pb.getKey()); });
+
```

Comment:
You are right, the `ProcessBinding` -> `Infrastructure` reference is indirect via the `StreamBinding` on inputs and outputs.

I feel like we should be as defensive as possible and still do the check on those inputs/outputs.

Reply:
Additionally look [here](https://github.com/ExpediaGroup/stream-registry/blob/master/core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/StreamBindingService.java#L96). It's currently possible to delete a `StreamBinding` when there is a reference it from a `ProcessBinding`. This could lead to dodgy data and orphaned entities.